#2 - Add training of existing models/scratch models

bjtulinl opened the pull request:
- Add training of existing ner and pos models

bjtulinl committed:
#2 - Add training of existing models/scratch models

- Add training of existing ner and pos models

jcklie commented:
Why do you not create a Corpus on the fly instead of writing to a file and loading it? There is `TaggedCorpus` which just needs tagged flair sentences.

bjtulinl committed:
#2 - Add training of existing models/scratch models

- Create the Corpus on the fly instead of writing and loading a file

jcklie commented:
What I mean is:
1. You create flair tokens for every sentence: https://github.com/zalandoresearch/flair/blob/master/flair/data.py#L170
2. You add these tokens to sentences: https://github.com/zalandoresearch/flair/blob/master/flair/data.py#L435
3. You add create a sentence dataset https://github.com/zalandoresearch/flair/blob/477d4004b4029efdd5af140bbb928101c305033e/flair/datasets.py#L356
4. You add these to a corpus: https://github.com/zalandoresearch/flair/blob/master/flair/data.py#L747

That is what flair internally is doing when you load a file

jcklie commented on app_flair.py:
```diff
@@ -96,6 +123,103 @@ def parse_prediction_request(json_object: JsonDict) -> PredictionRequest:
+    l = len(list_sentences)
+    test_size: int = round(l / 10)
+    dev_size: int = round(l / 10)
+    random.shuffle(list_sentences)
```
I think you are shuffling the wrong list here, you should not mutate input parameters. You can also just use sklearn: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

jcklie replied:
@bjtulinl Did you have a look at this?

bjtulinl replied:
Yes. But it seems that the function train_test_split in sklearn is used to split the dataset with the labels. In flair they use random_split in torch. So I write the function like this:
```
def dataset_split(list_sentences: list):
    l = len(list_sentences)
    dev_size: int = round(l / 10)
    train, dev = random_split(list_sentences,[l-dev_size,dev_size])
    train_l = len(train)
    test_size: int = round(train_l / 10)
    train, test = random_split(train, [train_l-test_size,test_size])
    return train, dev, test
```
Does it make sense?

jcklie replied:
in sklearn you do not need to the labels. Your split is difficult to read, as you split from train twice, making it not a 80 10 10 split and checking the indices is quite hard.

bjtulinl replied:
Then maybe:
def dataset_split(list_sentences: list):
    l = len(list_sentences)
    dev_size: int = round(l / 10)
    test_size: int = round(l / 10)
    train, dev = random_split(list_sentences,[l-dev_size,dev_size])
    train, test = random_split(train, [train_l-test_size,test_size])
    return train, dev, test
Or do you think sklearn is a better choice?

bjtulinl replied:
Then maybe:
def dataset_split(list_sentences: list):
l = len(list_sentences)
dev_size: int = round(l / 10)
test_size: int = round(l / 10)
train, dev = random_split(list_sentences,[l-dev_size,dev_size])
train, test = random_split(train, [I-dev_size-test_size,test_size])
return train, dev, test
Or do you think sklearn is a better choice?

jcklie replied:
I just would use sklearn. Do you really need the test set in flair?

bjtulinl replied:
No. We can just set the corpus as:
    corpus = Corpus( train=train , dev=dev, test='')
"train" is used to train the model and "dev" is used to evaluate the score of the model. And the learning rate is only changed  when the f1 score of dev set decreases.

jcklie replied:
Then I would just use either torch or sklearn split once and ignore the test set. That should make it simpler.

jcklie commented:
@bjtulinl Can you fix this merge conflict?

bjtulinl commented:
Ok. I will do that.
 
> On 1. Aug 2019, at 17:09, Jan-Christoph Klie <notifications@github.com> wrote:
> 
> @bjtulinl <https://github.com/bjtulinl> Can you fix this merge conflict?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/inception-project/external-recommender-flair/pull/16?email_source=notifications&email_token=AKVU7NQQNJO6YLLO24CGLQ3QCL4DFA5CNFSM4IH4LJQKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3K5DRA#issuecomment-517329348>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AKVU7NTKGVZERDSHV6OV7N3QCL4DFANCNFSM4IH4LJQA>.
>

bjtulinl committed:
Merge branch 'master' into feature/2-Add-training-of-existing-models/scratch-models

bjtulinl committed:
#2 - Add training of existing models/scratch models

- Update the split of dataset
- Remove redundant conversions from iterator(cas.select) to list
- Train the models without test sets

jcklie commented:
@bjtulinl It trains, but I am unable to get recommendations after the training. Does it work for you? I assume we need #20 .

bjtulinl commented:
I will check it.

> On 7. Aug 2019, at 17:14, Jan-Christoph Klie <notifications@github.com> wrote:
> 
> @bjtulinl <https://github.com/bjtulinl> It trains, but I am unable to get recommendations after the training. Does it work for you? I assume we need #20 <https://github.com/inception-project/external-recommender-flair/issues/20> .
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/inception-project/external-recommender-flair/pull/16?email_source=notifications&email_token=AKVU7NU4ZWNN4VKVNRXSLD3QDLRGJA5CNFSM4IH4LJQKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3YXXOY#issuecomment-519142331>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AKVU7NTNCOMZXN7Q3ES5BY3QDLRGJANCNFSM4IH4LJQA>.
>

jcklie commented:
I merge this and we will continue in #20 .
